modules/tensor_mechanics/src/kernels/TotalLagrangianStressDivergence.C

```diff
@@ -0,0 +1,354 @@
+#include "TotalLagrangianStressDivergence.h"
+
+#include "HomogenizationConstraintIntegral.h"
+
+registerMooseObject("TensorMechanicsApp", TotalLagrangianStressDivergence);
+
+InputParameters
+TotalLagrangianStressDivergence::validParams()
+{
+  InputParameters params = Kernel::validParams();
+
+  params.addRequiredParam<unsigned int>("component", "Which direction this kernel acts in");
+  params.addRequiredCoupledVar("displacements", "The displacement components");
+  params.addParam<bool>("kernel_large_kinematics", false, "Use large displacement kinematics");
+
+  // This kernel requires use_displaced_mesh to be off
+  params.suppressParameter<bool>("use_displaced_mesh");
+
+  params.addCoupledVar("macro_gradient", "Optional scalar field with the macro gradient");
+  params.addParam<std::vector<std::string>>("constraint_types",
+                                            "Type of each constraint: strain or stress");
+
+  return params;
+}
+
+TotalLagrangianStressDivergence::TotalLagrangianStressDivergence(const InputParameters & parameters)
+  : DerivativeMaterialInterface<Kernel>(parameters),
+    _ld(getParam<bool>("kernel_large_kinematics")),
+    _component(getParam<unsigned int>("component")),
+    _ndisp(coupledComponents("displacements")),
+    _disp_nums(_ndisp),
+    _disp_vars(_ndisp),
+    _grad_disp(_ndisp),
+    _stress(getMaterialPropertyByName<RankTwoTensor>("cauchy_stress")),
+    _material_jacobian(getMaterialPropertyByName<RankFourTensor>("material_jacobian")),
+    _inv_def_grad(getMaterialPropertyByName<RankTwoTensor>("inv_def_grad")),
+    _detJ(getMaterialPropertyByName<Real>("detJ")),
+    _df(getMaterialPropertyByName<RankTwoTensor>("inc_def_grad")),
+    _macro_gradient_num(isCoupledScalar("macro_gradient", 0) ? coupledScalar("macro_gradient") : 0),
+    _macro_gradient(isCoupledScalar("macro_gradient", 0) ? getScalarVar("macro_gradient", 0)
+                                                         : nullptr),
+    _indices(HomogenizationConstants::indices.at(_ld)[_ndisp - 1])
+{
+  if (isCoupledScalar("macro_gradient", 0))
+  {
+    // Check the order of the scalar variable
+    unsigned int needed = HomogenizationConstants::required.at(_ld)[_ndisp - 1];
+    if (_macro_gradient->order() != needed)
+      mooseError("The homogenization macro gradient variable must have order ", needed);
+
+    // Check the number of constraints
+    auto types = getParam<std::vector<std::string>>("constraint_types");
+    if (types.size() != needed)
+      mooseError("The kernel must be supplied ", needed, " constraint types.");
+  }
+}
+
+void
+TotalLagrangianStressDivergence::initialSetup()
+{
+  for (unsigned int i = 0; i < _ndisp; i++)
+  {
+    _disp_nums[i] = coupled("displacements", i);
+    _disp_vars[i] = getVar("displacements", i);
+    _grad_disp[i] = &coupledGradient("displacements", i);
+  }
+
+  if (isCoupledScalar("macro_gradient", 0))
+  {
+    auto types = getParam<std::vector<std::string>>("constraint_types");
```

Comment:
parameter type should be std::vector<MooseEnum> (this will also enable teh appropriate auto completion and allow you to remove the `map_string` method